De-dup Orleans unit testing content. (#39185)
dotnet/docs: docs/orleans/implementation/testing.md

## Use the `TestCluster`

The `Microsoft.Orleans.TestingHost` NuGet package contains <xref:Orleans.TestingHost.TestCluster> which can be used to create an in-memory cluster, comprised of two silos by default, which can be used to test grains.

:::code source="snippets/testing/Orleans-testing/Sample.OrleansTesting/HelloGrainTests.cs":::

Due to the overhead of starting an in-memory cluster, you may wish to create a `TestCluster` and reuse it among multiple test cases. For example, this can be done using xUnit's class or collection fixtures.

To share a `TestCluster` between multiple test cases, first create a fixture type:

:::code source="snippets/testing/Orleans-testing/Sample.OrleansTesting/ClusterFixture.cs":::

Next, create a collection fixture:

:::code source="snippets/testing/Orleans-testing/Sample.OrleansTesting/ClusterCollection.cs":::

You can now reuse a `TestCluster` in your test cases:

:::code source="snippets/testing/Orleans-testing/Sample.OrleansTesting/HelloGrainTestsWithFixture.cs":::

> xUnit will call the <xref:System.IDisposable.Dispose> method of the `ClusterFixture` type when all tests have been completed and the in-memory cluster silos are stopped. `TestCluster` also has a constructor that accepts <xref:Orleans.TestingHost.TestClusterOptions> that can be used to configure the silos in the cluster.

Reviewer: ```suggestion
xUnit calls the <xref:System.IDisposable.Dispose> method of the `ClusterFixture` type when all tests have been completed and the in-memory cluster silos are stopped. `TestCluster` also has a constructor that accepts <xref:Orleans.TestingHost.TestClusterOptions> that can be used to configure the silos in the cluster.
```

If you're using Dependency Injection in your Silo to make services available to Grains, you can use this pattern as well:

:::code source="snippets/testing/Orleans-testing/Sample.OrleansTesting/ClusterFixtureWithConfig.cs"

## Use mocks

Orleans also makes it possible to mock many parts of the system, and for many scenarios, this is the easiest way to unit test grains. This approach does have limitations (for example, around scheduling reentrancy and serialization) and may require that grains include code used only by your unit tests. The [Orleans TestKit](https://github.com/OrleansContrib/OrleansTestKit) provides an alternative approach, which side-steps many of these limitations.

> For example, let us imagine that the grain we're testing interacts with other grains. To be able to mock those other grains, we also need to mock the <xref:Orleans.Grain.GrainFactory> member of the grain under test. By default `GrainFactory` is a normal `protected` property, but most mocking frameworks require properties to be `public` and `virtual` to be able to mock them. So the first thing we need to do is make `GrainFactory` both `public` and `virtual` property:

Reviewer: ```suggestion
For example, imagine that the grain you're testing interacts with other grains. To be able to mock those other grains, you also need to mock the <xref:Orleans.Grain.GrainFactory> member of the grain under test. By default `GrainFactory` is a normal `protected` property, but most mocking frameworks require properties to be `public` and `virtual` to be able to mock them. So the first thing you need to do is make `GrainFactory` both a `public` and `virtual` property:
```

> Now we can create our grain outside of the Orleans runtime and use mocking to control the behavior of `GrainFactory`:

Reviewer: ```suggestion
Now you can create your grain outside of the Orleans runtime and use mocking to control the behavior of `GrainFactory`:
```

```csharp
using Xunit;
using Moq;

namespace Tests;

public class WorkerGrainTests
{
    [Fact]
    public async Task RecordsMessageInJournal()
    {
        var data = "Hello, World";
        var journal = new Mock<IJournalGrain>();
        var worker = new Mock<WorkerGrain>();
        worker
            .Setup(x => x.GrainFactory.GetGrain<IJournalGrain>(It.IsAny<Guid>()))
            .Returns(journal.Object);

        await worker.DoWork(data)

        journal.Verify(x => x.Record(data), Times.Once());
    }
}
```

> Here we create our grain under test, `WorkerGrain`, using Moq that means we can then override the behavior of the `GrainFactory` so that it returns a mocked `IJournalGrain`. We can then verify that our `WorkerGrain` interacts with the `IJournalGrain` as we expect.

Reviewer: ```suggestion
Here you create the grain under test, `WorkerGrain`, using Moq, which means you can override the behavior of the `GrainFactory` so that it returns a mocked `IJournalGrain`. You can then verify that the `WorkerGrain` interacts with the `IJournalGrain` as you expect.
```
